### ISSUE TITLE
Request to import pytest in tests/*.py

### ISSUE THREAD

**Issue by fritzo** [pytorch/pytorch#11578]
Currently PyTorch uses the builtin `unittest` module for testing. Would it be possible to add a dependency on pytest so developers can more easily write parametrized tests?

While working on tests/test_distributions.py, @neerajprad, @alicanb, and @fritzo have found it challenging to write heavily parametrized tests. As a result, test coverage suffers. By contrast in Pyro we use heavily parametrized tests using pytest, and our Pyro tests seem to catch many bugs that aren't caught in PyTorch's own tests (#9917, #9521, #9977, #10241).

In particular, I'd like to be able to gather xfailing tests in batch
```py
DISTRIBUTIONS = [Bernoulli, Beta, Cauchy, ..., Weibull]

@pytest.mark.parametrize('Dist', DISTRIBUTIONS)
def test_that_often_fails(Dist):
    dist = Dist(...)
    assert something(dist)
```
and then mark xfailing parameters
```
DISTRIBUTIONS = [
    Bernoulli,
    Beta,
    pytest.param(Cauchy, marks=[pytest.mark.xfail(reason='schema not found for node')]),
    ...
    Weibull,
]
```
pytest makes it easy to collect xfailing tests in batch and to see the entire list of xfailing tests, and makes it easy to run xfailing tests to see which have started passing since the last time tests were run. In contrast, `unittest` fixtures typically parametrize via for loops and can report only a single failure on each run.

I think allowing usage of pytest in tests/*.py could help improve PyTorch test coverage.

**Comment by fritzo** [comment]
cc @neerajprad @vishwakftw @alicanb

**Comment by vishwakftw** [comment]
As a precursor, we could try porting all tests in `test_distributions.py` from `unittest` to `pytest` (probtorch/pytorch, perhaps) to provide a working comparison.

**Comment by neerajprad** [comment]
I think the ease of writing, debugging and marking (for skipping/xfailing) parameterized tests using pytest may also be useful in other parts of the codebase like `test_jit.py`.

**Comment by zou3519** [comment]
@apaszke had some thoughts about this from a few months back

**Comment by fritzo** [comment]
@ezyang what are your thoughts on allowing `import pytest`?

I'd like to move more Pyro distributions upstream to PyTorch so we can share code with @nimar and @jpchen and team inside Facebook. However I'm reluctant to move that code unless we can move the tests mostly intact, and those tests heavily use `pytest.mark.parametrize`.

**Comment by ezyang** [comment]
Since we have hypothesis in the test suite now, my judgment is that I think pytest should be fine.

**Comment by mruberry** [comment]
What happens when these tests are run using unittest, though? Do they run as expected?

**Comment by fritzo** [comment]
@mruberry Oh good point, `python -m unittest` does not collect pytest-style tests, so we would need to switch to the `pytest` runner at least for test_distributions.py

https://github.com/pytorch/pytorch/blob/4c99a9b6721c445a8c427020e54b63da0dbe622c/CONTRIBUTING.md#L215-L223

What would it take to officially support pytest as the test runner? Would that require reconfiguring facebook internal systems, or could that be accomplished by a community contributor like me?

**Comment by ezyang** [comment]
I am not quite sure I understand your question. Are you saying that, after converting the tests to use pytest features, it would be *mandatory* to run `pytest` to run these tests? (E.g., we couldn't do something like https://www.avivroth.com/2018/03/06/python-embedding-pytest-in-code/ to ensure that `python test/test_nn.py` kept working?) If the runner mechanism changes completely, it wouldn't just be FB internal you'd have to update; you'd also have to fix all of our CI and packaging scripts to invoke test scripts in the new. I would deeply, deeply prefer if the old idiom of just calling `python test/test_nn.py` kept working.

I checked our internal status on pytest; according to https://fb.workplace.com/groups/fbpython/permalink/3636251993082808/ (FB-only, sorry) we only recently (March) got support for pytest style tests in fbcode. So hypothetically a pytest runner would work there, but it would definitely need work from FB side.

By the way, I have some experience using pytest from the onnx project, and I am not a big fan of some of pytest's defaults; for example, I hate how it dumps all warnings at the end of stderr after the errors (show me the errors last!) and I also hate the pytest error reporting format, which has a huge verbose source code dump and backtrace that isn't really that useful. If we switch to pytest, it would be good also spend some time going through the configuration knobs and making sure we don't configure things in a bad way.

cc @malfet @kostmo

**Comment by malfet** [comment]
To the best of my knowledge, pytest should be compatible with most of the existing unittest tests, for example I can run almost all the tests from quantization folder:
```
$ pushd test/quantization; pytest -v ; popd
~/git/pytorch/test/quantization ~/git/pytorch
==================================================================================================================== test session starts =====================================================================================================================
platform linux -- Python 3.7.7, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /usr/local/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/nshulga/git/pytorch/test/quantization/.hypothesis/examples')
rootdir: /home/nshulga/git/pytorch
plugins: hypothesis-5.8.0
collected 221 items                                                                                                                                                                                                                                          

test_backward_compatibility.py::TestSerialization::test_conv2d PASSED                                                                                                                                                                                  [  0%]
test_backward_compatibility.py::TestSerialization::test_conv2d_relu PASSED                                                                                                                                                                             [  0%]
test_backward_compatibility.py::TestSerialization::test_conv3d PASSED                                                                                                                                                                                  [  1%]
test_backward_compatibility.py::TestSerialization::test_conv3d_relu PASSED                                                                                                                                                                             [  1%]
test_backward_compatibility.py::TestSerialization::test_linear PASSED                                                                                                                                                                                  [  2%]
test_backward_compatibility.py::TestSerialization::test_linear_dynamic PASSED                                                                                                                                                                          [  2%]
test_backward_compatibility.py::TestSerialization::test_linear_relu PASSED                                                                                                                                                                             [  3%]
test_numeric_suite.py::TestEagerModeNumericSuite::test_compare_model_outputs PASSED                                                                                                                                                                    [  3%]
test_numeric_suite.py::TestEagerModeNumericSuite::test_compare_model_stub PASSED                                                                                                                                                                       [  4%]
test_numeric_suite.py::TestEagerModeNumericSuite::test_compare_weights PASSED                                                                                                                                                                          [  4%]
test_qat_module.py::TestQATModule::test_conv_bn_relu PASSED                                                                                                                                                                                            [  4%]
test_quantize.py::TestPostTrainingStatic::test_activations PASSED                                                                                                                                                                                      [  5%]
test_quantize.py::TestPostTrainingStatic::test_manual PASSED                                                                                                                                                                                           [  5%]
test_quantize.py::TestPostTrainingStatic::test_nested1 PASSED                                                                                                                                                                                          [  6%]
test_quantize.py::TestPostTrainingStatic::test_nested2 PASSED                                                                                                                                                                                          [  6%]
test_quantize.py::TestPostTrainingStatic::test_nested3 PASSED                                                                                                                                                                                          [  7%]
test_quantize.py::TestPostTrainingStatic::test_normalization PASSED                                                                                                                                                                                    [  7%]
test_quantize.py::TestPostTrainingStatic::test_resnet_base PASSED                                                                                                                                                                                      [  8%]
test_quantize.py::TestPostTrainingStatic::test_save_load_state_dict PASSED                                                                                                                                                                             [  8%]
test_quantize.py::TestPostTrainingStatic::test_single_layer PASSED                                                                                                                                                                                     [  9%]
test_quantize.py::TestPostTrainingStatic::test_skip_quant PASSED                                                                                                                                                                                       [  9%]
test_quantize.py::TestPostTrainingStatic::test_two_layers PASSED                                                                                                                                                                                       [  9%]
test_quantize.py::TestQuantizationAwareTraining::test_activations PASSED                                                                                                                                                                               [ 10%]
test_quantize.py::TestQuantizationAwareTraining::test_conv_linear PASSED                                                                                                                                                                               [ 10%]
test_quantize.py::TestQuantizationAwareTraining::test_eval_only_fake_quant PASSED                                                                                                                                                                      [ 11%]
test_quantize.py::TestQuantizationAwareTraining::test_manual PASSED                                                                                                                                                                                    [ 11%]
test_quantize.py::TestQuantizationAwareTraining::test_train_save_load_eval PASSED                                                                                                                                                                      [ 12%]
test_quantize.py::TestGraphModePostTrainingStatic::test_conv PASSED                                                                                                                                                                                    [ 12%]
test_quantize.py::TestGraphModePostTrainingStatic::test_conv_bn SKIPPED                                                                                                                                                                                [ 13%]
test_quantize.py::TestGraphModePostTrainingStatic::test_nested PASSED                                                                                                                                                                                  [ 13%]
...
```

**Comment by mruberry** [comment]
We do try to keep our test suite working with pytest, but actually switching to it seems like a lot of lift. It would also prevent people who prefer using unittest from using it, at least on this test file.

What would you think about adding a similar decorator to the generic test framework instead? We, could, for example, add a @parameterize decorator that accepted an iterable of strings, then you could lookup the parameters in a str->args dict:

```
@parameterize('add', 'sub')
def my_generic_test(self, device, param):
  arg_a, arg_b ... = my_generic_test_params[param]
  ...
```

This would generate tests like:

```
my_generic_test_cpu_add
my_generic_test_cpu_sub
my_generic_test_cuda_add
my_generic_test_cuda_sub
```

This might get you 80% of what you want.

**Comment by fritzo** [comment]
Thanks for looking into this!

> @ezyang
> ... do something like [embedding](https://www.avivroth.com/2018/03/06/python-embedding-pytest-in-code) to ensure that `python test/test_nn.py` kept working?
> ```py
> if __name__ == "__main__":
>     pytest.run([__file__])
> ```
> I would ... prefer if the old idiom of just calling `python test/test_nn.py` kept working

:+1: I wasn't aware of this solution, and it seems like an ideal solution from my perspective: it would preserve your `python test.py` idiom and would support all pytest features used in existing tests I'd like to port. I think you're right that the unittest runner would not be supported for this file `python -m unittest test_distributions.py`.

Would you be ok if we **(1)** split up `test/test_distributions.py` into multiple files `test/distributions/*` and **(2)** use pytest and embedding in that directory (taking care to configure defaults to your liking). I could also try to get `python -m unittest ...` working or at least failing with an error like "skipped pytest-style tests" to avoid silent failure.

> @mruberry 
> adding a similar decorator ... might get you 80% of what you want.

I could see `@parametrize` working if we exactly match the syntax of `@pytest.mark.parametrize` and thus minimize the friction of migrating existing tests. If @ezyang's embedding solution is infeasible, I bet we could match existing syntax, as in this [distributions test](https://github.com/pyro-ppl/pyro/blob/cce694178ae66b3a84623d517ffdf0c1bb32ba27/tests/distributions/test_hmm.py#L157):
```py
@pytest.mark.parametrize('event_shape', [(), (5,), (2, 3)], ids=str)
@pytest.mark.parametrize('state_dim', [2, 3])
@pytest.mark.parametrize('num_steps', [1, 2, 3])
@pytest.mark.parametrize('init_shape,trans_shape,obs_shape', [
    ((), (), ()),
    ((), (1,), ()),
    ((), (), (1,)),
    ((), (1,), (7, 1)),
    ((), (7, 1), (1,)),
    ((), (7, 1), (7, 1)),
    ((7,), (1,), (1,)),
    ((7,), (1,), (7, 1)),
    ((7,), (7, 1), (1,)),
    ((7,), (7, 1), (7, 1)),
    ((4, 1, 1), (3, 1, 1), (2, 1)),
], ids=str)
def test_discrete_hmm_homogeneous_trick(init_shape, trans_shape, obs_shape, event_shape, state_dim, num_steps):
    ...
```

> @malfet
> pytest should be compatible with most of the existing unittest tests

Yeah I think many PyTorch devs already use pytest as their runner. I think the concern is that new pytest-style test code wouldn't be runnable with the older `python -m unittest test_file.py` (but as @ezyang points out we could still run with `python test_file.py`).

**Comment by mruberry** [comment]
Cool, seems like we're narrowing down to a couple reasonable-looking options. 

If we did make our own parameterize it would probably only apply to our generic tests. Are you familiar with that framework, and what do you think about moving test_distributions to it? It's a great way to generate device and dtype test variants, so this functionality would be a natural extension for it.

**Comment by mthrok** [comment]
Domain library is also looking for a solution to test parameterization. torchaudio has some function-style tests written with `pytest.mark.parametrize` and I am tying to find the best way to migrate it so that it can run in fbcode with unittest runner. (yet we can use `pytest` as runner on Circle CI since `pytest` as runner gives better messages, in my opinion)

Either custom `@parameterize` decorator or third party solution [`@parameterized.expand`](https://pypi.org/project/parameterized/) would be sufficient for my case, but if PyTorch has solution on it, I think domain libraries can align to it.

**Comment by neerajprad** [comment]
@ezyang - Adding pytest will make it really easy for us to port tests from Pyro, and should involve no changes in CI in the way tests are invoked (as you pointed out) except that we will need to configure the CI scripts to install pytest. Based on the discussion above, am I right in assuming that we are in principle fine with adding a dependency (test time) on pytest, since we will need to do that if we are to use pytest as more than a test runner, e.g. using parameterized tests, fixtures, conditional skipping, etc? Note that this will only affect new distribution tests that use some pytest feature and all remaining tests should run with both unittest and pytest. We can also mark these files that import pytest so that pytorch developers who don't have pytest installed are simply able to skip these tests when they run `run_test.py`. Does this seem reasonable?

**Comment by ezyang** [comment]
Yes, if you can figure out how to do this without breaking all of the external-to-Python fixtures (besides needing to install pytest), you have my blessing. You don't need to even skip, we can just force people to install pytest. And even if you can't get it to work, please holler, we probably can figure something out.

**Comment by mthrok** [comment]
I think that we can start from turning `assertEqual` method of `torch.testing._internal.common_utils.TestCase` into a plain function. (Not sure if it's already done.) The existing method can just pass the arguments to functional form.
This is actually enough for domain libraries to convert the tests into functional style and fully adopt pytest, because that's the only test method we use and we heavily rely on.

**Comment by neerajprad** [comment]
> I think that we can start from turning assertEqual method of torch.testing._internal.common_utils.TestCase into a plain function. (Not sure if it's already done.)

My suggestion would be to have something like [np.testing.assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) which lets us specify absolute vs relative precision and gives very nice error messages (see below) that can be shared by all test runners:

```
AssertionError:
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 1 / 3 (33.3%)
Max absolute difference: 1.e-05
Max relative difference: 4.999975e-06
 x: array([0., 1., 2.])
 y: array([0.     , 1.     , 2.00001])
``` 

Even the standard python assert statement gives good error messages and stack trace in pytest but we will probably want something more detailed when checking for tensor equality.

**Comment by rgommers** [comment]
Here is a summary of the current status:

- It's already possible to use pytest features in new test files. There are unconditional `import pytest`'s in a number of files in `test/distributions/` and `test/distributed/pipeline/sync/`.
- Such files are listed in `test/run_test.py` in `USE_PYTEST_LIST`. If `pytest` is not installed, a fallback to run those tests with `unittest` is tried (which may or may not work of course).
- It's fine to make `pytest` an unconditional dependency, just like `hypothesis` (comment @ezyang above: https://github.com/pytorch/pytorch/issues/11578#issuecomment-701712308). This is not done.
- When making `pytest` an unconditional dependency, it's necessary to keep the workflow of running `python test/run_test.py` (or an individual file, e.g. `python test/test_nn.py`) unchanged. Which is possible. Keeping `python -m unittest test/test_nn.py` does not need to remain working.
- `pytest`'s default output is too verbose and annoying, that should be addressed: make it not print docstrings (gh-3791), and use short tracebacks rather than the default long ones.

There's one comment I don't quite understand, missing some context on what the "external-to_Python fixtures" are exactly:

> Yes, if you can figure out how to do this without breaking all of the external-to-Python fixtures (besides needing to install pytest), you have my blessing. You don't need to even skip, we can just force people to install pytest. And even if you can't get it to work, please holler, we probably can figure something out.

There are a number of advantages to switching to pytest, some of which are mentioned above:

- Using modern test features that `unittest` doesn't have:

  - `parametrize`
  - `fixtures`
  - collect xfailing tests in batch
  - more control on which tests to run, e.g. using a regex
  - using test functions and plain asserts rather than the more clunky "must define a class and use `assertTrue` and other `TestCase` methods
  - easier to run tests in parallel locally with `pytest-xdist` (this may result in occasional flakes because the test suite is currently not parallel-safe, but it's very useful for local development)
 
- Much easier to port code from other libraries, such as from `pyro`. We also run into this when reusing code from NumPy and SciPy. (e.g., gh-52408 "Port NumPy typing testing style to PyTorch" is blocked on this).
- It reduces friction. Almost every large Python package uses `pytest`, and many devs are used to this. Having to use `unittest` which is missing lots of features and has more verbose syntax feels annoying and unproductive (even if the actual time loss may be limited in a lot of cases)
- Make it easier to implement other CI improvements, like turning deprecating warnings into errors (gh-49642)

Many people already use pytest as a test runner, so improving the traceback format is a good start (easy to do by adding a `pytest.ini` file). I'll open a new tracking issue with a more detailed plan (on request of @mruberry).

**Comment by NicolasHug** [comment]
Some of the most useful `pytest` features like parametrization and fixtures are unfortunately [incompatible with `unittest.TestCase` methods](https://docs.pytest.org/en/stable/unittest.html#pytest-features-in-unittest-testcase-subclasses), which I believe account for most of PyTorch tests (I could be wrong on this, but at least that's the case in `torchaudio` and `torchvision`). This means that for now, only newly written tests could be parametrized and use `pytest` fixtures.

Is there a plan to port the existing tests relying on `unittest.TestCase` into functions that would be compatible with `@pytest.parametrize`?

> My suggestion would be to have something like np.testing.assert_allclose

I'd be happy to give this a shot

**Comment by mruberry** [comment]
> Is there a plan to port the existing tests relying on `unittest.TestCase` into functions that would be compatible with `@pytest.parametrize`?

No, but we don't need to add the ability to parameterize existing tests. We also override methods like assertEqual() today, so another option might be to not call the unittest methods from our test class(es).

> My suggestion would be to have something like np.testing.assert_allclose

It would be nice to have a torch.testing module, ideally one compatible with NumPy's testing module, too, but we should think about the design a bit before getting started.

**Comment by jbschlosser** [comment]
FYI update on this: I've added a test parametrization mechanism that *is* compatible with the testing infrastructure in PyTorch core. It was designed to operate in a very similar way to `@pytest.parametrize`. See the `@parametrize` decorator defined [here](https://github.com/pytorch/pytorch/blob/b7742b437accdc1068f739c16f685970537f158a/torch/testing/_internal/common_utils.py#L229-L279) for documentation. Example of real usage in the wild:https://github.com/pytorch/pytorch/blob/b7742b437accdc1068f739c16f685970537f158a/test/test_autograd.py#L4313-L4315

Note that instances of `@parametrize` stack via the product of args (just like pytest).

Tiny caveat: `test/test_nn.py` imports `@parametrize` as `@parametrize_test` to avoid conflicting with the unrelated [parametrization mechanism for torch.nn modules](https://pytorch.org/tutorials/intermediate/parametrizations.html).
